### PR TITLE
Fix "var x = nil;"

### LINKED ISSUE #19244: Internal error when assigning c_nil to untyped var declaration
### Summary of Problem

I'm getting an internal error when assigning `c_nil` to an untyped `var` declaration.

### Steps to Reproduce

**Source Code:**

```chapel
use CPtr;

var ptr = c_nil;
```

**Compile command:**
```
~/bugs> chpl test.chpl
test.chpl:3: internal error: AST-AGG-YPE-2003 chpl version 1.25.1
Note: This source location is a guess.

Internal errors indicate a bug in the Chapel compiler ("It's us, not you"),
and we're sorry for the hassle.  We would appreciate your reporting this bug -- 
please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
the filename + line number above may be useful in working around the issue.

~/bugs> chpl test.chpl --devel
test.chpl:3: internal error: no field 'nil' in class 'heap__nilType' in getField() [AST/AggregateType.cpp:2003]
Note: This source location is a guess.
```

### Configuration Information
```
chpl version 1.25.1
  built with LLVM version 11.0.1
Copyright 2020-2021 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)

CHPL_TARGET_PLATFORM: cray-xc
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: broadwell
CHPL_LOCALE_MODEL: flat
CHPL_COMM: ugni
CHPL_TASKS: qthreads
CHPL_LAUNCHER: aprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: ugni
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: bundled
CHPL_AUX_FILESYS: none
```

### PR DESCRIPTION
**Pull request by vasslitvinov**
Resolves #19244

This changes enables the compiler to compile code like this:
`var x1 = nil;`
`var x2 = c_nil;`

which used to cause assertion failures during --no-local compilations.

Note that #19672 proposes to disallow this code.
That will likely make this fix unnecessary.

While there:

* Enable nil-variable-elimination upon --local.
  Previously it was done only during --no-local compilations.

* Resolve, unintentionally, a gasnet --verify failure in
  `test/library/packages/canCompileNoLink/HDFStest.chpl`

Testing: standard paratest; multilocale tests under gasnet.

### CODE REVIEW

File: test/classes/vass/var_x_=_nil.chpl [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+var x = nil;
+writeln(x);
+writeln(ptr);
+
+record RR {
+  var value = nil;
+}
+
+var rr: RR;
+__primitive("=", rr.value, nil); // this happens upon "rr.value = nil;"
```
Comment (lydia-duncan) [inline test/classes/vass/var_x_=_nil.chpl:20]: Why is this using the primitive instead of `rr.value = nil` itself?

**Review by lydia-duncan — APPROVED** [review]
Just a small question, otherwise this looks good to me!

File: test/classes/vass/var_x_=_nil.chpl (hunk 1 above)
Comment (vasslitvinov) [reply test/classes/vass/var_x_=_nil.chpl:20]: There are two overloads of `operator =` in ChapelBase.chpl that match `nil = nil;` -- one for borrowed nilable classes and the other for unmanaged nilable classes. So writing `rr.value = nil;` gave me an "ambiguous call" error. Inlining the body of either overload allowed the test to compile.

Comment (lydia-duncan) [reply test/classes/vass/var_x_=_nil.chpl:20]: Huh.  That's unfortunate, but makes sense

**Pull request merged**